# Mutational Frequency in larger cohorts

In this section we check the mutational frequency of common drivers associated with DCIS in a IDC cohort. This includes the common oncogenes TP53, PIK3CA, GATA3, PTEN but also includes commonly mutated genes including TTN and MUC16. 

## Mutational frequency in TCGA

```{r}
B1$TumName=substr(B1$Tumor_Sample_Barcode, 1, 12)

mSearch=c("AKT1","CDH1",  "GATA3", "KMT2C", "MAP3K1" ,"MUC16" , "PALB2", "PIK3CA", "PTEN" ,  "TP53" ,  "TTN" ) # add PTEN, KMT2C
lx1=which(B1$Hugo_Symbol%in%mSearch)
B2=B1[lx1, ]

```


Below is a summary of the mutational and neoantigen frequency of these genes in the TCGA cohort. The most common PIK3CA mutatiions are at amino acids 1047, 545, 542, 345, 726, 453 (more than 5 cases with mutations). 
In TP53, these sites are 220 248 193 273 175. 

```{r}
## Loop over all genes

AllGMat=matrix(NA, nrow=length(namList), ncol=12)

colnames(AllGMat)=c("N.IDC.mutations",  "N.IDC.neoant",  "N.IDC.rsSNPs",  "N.IDC.uniqueSites", "N.IDC.sitesNeo", "N.IDC.sitesdbSNP", "PropNeoTotal", "PropSitesNeo", "PropdbSNP", "N.IDC.patmut","N.IDC.patNeo","N.IDC.pat.rsSNP")
rownames(AllGMat)=namList

CommSites=rep(NA, length(namList))

B2new=data.frame()

for (i in 1:length(allTCGANeoList2)){
  allTCGANeoList2[[i]]$samp2=substr(allTCGANeoList2[[i]]$barcode, 1, 12)
  
  ## get the matrix with gene of interest
  Bgata=B2[which(B2$Hugo_Symbol==namList[i]), c("Hugo_Symbol", "Chromosome", "Start_position","Protein_Change", "TumName", "dbSNP_RS", "i_tumor_vaf")]
  Bgata=unique(Bgata)
  AllGMat[i, 1]=nrow(Bgata) ## or length(unique(Bgata$TumName))
  AllGMat[i, 10]=length(unique(Bgata$TumName))
  ## redo this part
  Bgata$Neoant=NA
  p1=paste(Bgata$TumName, Bgata$Start_position)
  g1=paste(allTCGANeoList2[[i]]$samp2, allTCGANeoList2[[i]]$position)
  
  Bgata$Neoant[which(p1%in%g1)]=1
  AllGMat[i, 2]=sum(Bgata$Neoant, na.rm=T) 
  AllGMat[i, 11]=sum(sign(table(Bgata$TumName, Bgata$Neoant)))
  AllGMat[i, 3]=length(which(Bgata$dbSNP_RS!="")) 
  Bgata$rsSNP2=ifelse(Bgata$dbSNP_RS=="", NA, 1)
  AllGMat[i, 12]=sum(sign(table(Bgata$TumName, Bgata$rsSNP2)))
  
  lx1=regmatches(as.character(Bgata$Protein_Change), regexpr("[0-9]+", as.character(Bgata$Protein_Change)))
  Bgata$Protein_Change2=NA
  Bgata$Protein_Change2[which(Bgata$Protein_Change!="")]=lx1
  
  a1=table(factor(Bgata$Protein_Change2[which(Bgata$Neoant==1)], levels=unique(Bgata$Protein_Change2)))
  a2=table(factor(Bgata$Protein_Change2, levels=unique(Bgata$Protein_Change2)))
  a3=table(factor(Bgata$Protein_Change2[which(Bgata$dbSNP_RS!="")], levels=unique(Bgata$Protein_Change2)))
  
  AllGMat[i, 4]=length(a1)
  
  AllGMat[i, 5]=length(which(a1>0))
  AllGMat[i, 6]=length(which(a3>0))
  AllGMat[i, 7]=sum(a1)/sum(a2)
  AllGMat[i, 8]=mean(a1/a2)
  AllGMat[i, 9]=sum(a3)/sum(a2)
  
  P2=table(factor(substr(Bgata$Protein_Change, 3, 10)))
  CommSites[i]=paste(names(which(P2>ceiling(as.numeric(AllGMat[i, 4])/20))),collapse =", " )
  B2new=rbind(B2new, Bgata)
}

AllGMat
```

Visualising the frequency in the patient population below. Note that some patients may have multiple mutations in the same gene, hence, the total mutation count appears to be higher than the patient count. 

In red are the proportion of patients with a mutation, green indicates the proportions of patients with mutation predicted to be a neoantigen. Blue indicates whether this is a hotspot or rsSNP site. PIK3CA for example sees that most of its mutations are neoantigens (approximately 90\%) and they almost all occur at annotated rsSNP sites.


```{r}

anew=melt(AllGMat[ ,c(10:12)]/length(unique(B1$TumName)))
ggplot(anew, aes(x=Var1, y=value, fill=Var2))+geom_bar(stat="identity", position='dodge')+theme_bw()+xlab("gene")+ylab("frequency in patient population")
```

Note that TP53 and PIK3CA mutations below seem to be occurring at hostpots: whilst for the rest of the genes the number of unique sites at which mutations are found increases with patient frequency, TP53 and PIK3CA have fewer sites despite high frequency 

```{r, fig.height=6, fig.width=6}
plot(AllGMat[ ,10], AllGMat[ ,4], xlab="Patient frequency", ylab="No unique sites")
text(AllGMat[ ,10], AllGMat[ ,4], rownames(AllGMat))
```

## HLA neoantigen frequency

Summary of the HLA frequencies in the TCGA breast patients

```{r}
AllHLAs=read.delim("data/DontUpload/OptiTypeCallsHLA_20171207.tsv", sep=",")
AllHLAs$sub=substr(AllHLAs$aliquot_id, 1, 12)

midx=which(AllHLAs$sub%in%as.character(BrClin$Patient.ID))
AllHLAs=AllHLAs[midx, ]

Atable=table(unlist(AllHLAs[ ,1:2]))
Btable=table(unlist(AllHLAs[ ,3:4]))
Ctable=table(unlist(AllHLAs[ ,5:6]))

par(mfrow=c(3,1))

barplot(sort(Atable[which(Atable>5)]), las=2, ylab="frequency A allele")
mtext("TCGA common HLAs", 3)
barplot(sort(Btable[which(Btable>5)]), las=2, ylab="frequency B allele")
barplot(sort(Ctable[which(Ctable>5)]), las=2, ylab="frequency C allele")


```

```{r}
HLAlistTCGA=lapply(allTCGANeoList2, function(x) unique(x[ ,c("barcode", "hla", "position")]))

# number of times a particular HLA appears
UnHLAlist=lapply(HLAlistTCGA, function(x) table(x$hla))

# get the HLA-Avalues first

HSummA=lapply(UnHLAlist, function(x) x[grep("HLA-A", names(x))])
HSummB=lapply(UnHLAlist, function(x) x[grep("HLA-B", names(x))])
HSummC=lapply(UnHLAlist, function(x) x[grep("HLA-C", names(x))])

par(mfrow=c(2,2))

for (i in c( 6, 8, 10:11)){
  barplot(sort(HSummA[[i]][which(HSummA[[i]]>5)]), las=2, main=names(HSummA)[i], ylab="HLA-A frequency")
}

for (i in c( 6, 8, 10:11)){
  barplot(sort(HSummB[[i]][which(HSummB[[i]]>5)]), las=2, main=names(HSummB)[i], ylab="HLA-B frequency")
}

for (i in c( 6, 8, 10:11)){
  barplot(sort(HSummC[[i]][which(HSummC[[i]]>5)]), las=2, main=names(HSummC)[i], ylab="HLA-C frequency")
}

## For PIK3CA and TP53, check how frequently the mutations appear

TabPosHLA=lapply(HLAlistTCGA, function(x) table(x$position, x$hla))

PIK3CAtab=TabPosHLA[[8]]
coMatch=B2$Protein_Change[match(rownames(PIK3CAtab), B2$Start_position)]
rownames(PIK3CAtab)=coMatch
midx=rowSums(PIK3CAtab)
PIK3CAtab=PIK3CAtab[order(midx, decreasing = T), ]


TP53tab=TabPosHLA[[10]]
coMatch=B2$Protein_Change[match(rownames(TP53tab), B2$Start_position)]
rownames(TP53tab)=coMatch
midx=rowSums(TP53tab)
TP53tab=TP53tab[order(midx, decreasing = T), ]

rmIdx=which(colSums(PIK3CAtab)>50)
rmIDx2=which(rowSums(PIK3CAtab)>50)

HLAgrep=c("HLA-B*44:02", "HLA-A*11:01", "HLA-B*40:01", "HLA-C*07:02", "HLA-A*11:01", "HLA-A*33:03", "HLA-C*07:06", "HLA-B*58:01")

al1=match(HLAgrep, colnames(PIK3CAtab))

heatmap.2(PIK3CAtab[ rmIDx2,na.omit(unique(c(rmIdx, al1)))], col=brewer.pal(9, "Blues"), scale="none", trace="none")

rmIdx=which(colSums(TP53tab)>100)
rmIDx2=which(rowSums(TP53tab)>100)

HLAgrep=c("HLA-A*02:01", "HLA-A*31:01", "HLA-B*40:01", "HLA-A*33:03", "HLA-C*07:06", "HLA-C*03:03", "HLA-C*07:06", "HLA-A*68:01", "HLA-C*04:01")

al1=match(HLAgrep, colnames(TP53tab))
type2=c("p.G266R", "p.R273H", "p.P151T", "p.C176F", "p.R248W")
a12=match(type2, rownames(TP53tab))
heatmap.2(TP53tab[ na.omit(unique(c(rmIDx2, a12))),na.omit(unique(c(rmIdx, al1)))], col=brewer.pal(9, "Blues"), scale="none", trace="none")

```


## Association with immune signatures

Is there an association between samples which are predicted to be neoantigens and immune-function?

Below, TCGA cases are split into "mutated" vs "wild type" for the gene genes of interest. A t.test is performed to determine whether there is a difference in the Thorsson signatures between the mutant compared to the wild type population. 

Only associations with p <0.05 are shown. The heatmaps show the difference between the mean mutant signature and the mean wild type signature. (Red indicates higher in the mutant).

In the below analysis, we focus only on ER+ cases due to the higher number of patients, and to avoid the confounding effect of association of immune scores with patient subtype.

```{r}
lx1=acast(B2new[ ,c("TumName", "Hugo_Symbol")],TumName~Hugo_Symbol)
lx2=match(substr(rownames(lx1), 1, 12), ThorssonTab$`TCGA Participant Barcode`)

## mutations?

ThorssonTab$AKT1=0
ThorssonTab$AKT1[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"AKT1"]))
ThorssonTab$CDH1=0
ThorssonTab$CDH1[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"CDH1"]))
ThorssonTab$GATA3=0
ThorssonTab$GATA3[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"GATA3"]))
ThorssonTab$MUC16=0
ThorssonTab$MUC16[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"MUC16"]))
ThorssonTab$MAP3K1=0
ThorssonTab$MAP3K1[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"MAP3K1"]))
ThorssonTab$KMT2C=0
ThorssonTab$KMT2C[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"KMT2C"]))
ThorssonTab$PALB2=0
ThorssonTab$PALB2[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"PALB2"]))
ThorssonTab$PIK3CA=0
ThorssonTab$PIK3CA[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"PIK3CA"]))
ThorssonTab$PTEN=0
ThorssonTab$PTEN[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"PTEN"]))
ThorssonTab$TP53=0
ThorssonTab$TP53[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"TP53"]))
ThorssonTab$TTN=0
ThorssonTab$TTN[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"TTN"]))

## create a linear model here?

## mutations?
B2neo=B2new[which(B2new$Neoant==1), ]
lx1=acast(B2neo[ ,c("TumName", "Hugo_Symbol")],TumName~Hugo_Symbol)
lx2=match(substr(rownames(lx1), 1, 12), ThorssonTab$`TCGA Participant Barcode`)


ThorssonTab$AKT1n=0
ThorssonTab$AKT1n[(lx2)]=(sign(lx1[,"AKT1"]))
ThorssonTab$CDH1n=0
ThorssonTab$CDH1n[(lx2)]=(sign(lx1[ ,"CDH1"]))
ThorssonTab$GATA3n=0
ThorssonTab$GATA3n[(lx2)]=(sign(lx1[,"GATA3"]))
ThorssonTab$MUC16n=0
ThorssonTab$MUC16n[(lx2)]=(sign(lx1[ ,"MUC16"]))
ThorssonTab$MAP3K1n=0
ThorssonTab$MAP3K1n[(lx2)]=(sign(lx1[ ,"MAP3K1"]))
ThorssonTab$KMT2Cn=0
ThorssonTab$KMT2Cn[(lx2)]=(sign(lx1[ ,"KMT2C"]))
ThorssonTab$PALB2n=0
ThorssonTab$PALB2n[(lx2)]=(sign(lx1[ ,"PALB2"]))
ThorssonTab$PIK3CAn=0
ThorssonTab$PIK3CAn[(lx2)]=(sign(lx1[ ,"PIK3CA"]))
ThorssonTab$PTENn=0
ThorssonTab$PTENn[(lx2)]=(sign(lx1[ ,"PTEN"]))
ThorssonTab$TP53n=0
ThorssonTab$TP53n[(lx2)]=(sign(lx1[,"TP53"]))
ThorssonTab$TTNn=0
ThorssonTab$TTNn[(lx2)]=(sign(lx1[,"TTN"]))

## mutations?
B2rs=B2new[which(B2new$rsSNP2==1), ]
lx1=acast(B2rs[ ,c("TumName", "Hugo_Symbol")],TumName~Hugo_Symbol)
lx2=match(substr(rownames(lx1), 1, 12), ThorssonTab$`TCGA Participant Barcode`)

ThorssonTab$CDH1rs=0
ThorssonTab$CDH1rs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"CDH1"]))
ThorssonTab$GATA3rs=0
ThorssonTab$GATA3rs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"GATA3"]))
ThorssonTab$MUC16rs=0
ThorssonTab$MUC16rs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"MUC16"]))
ThorssonTab$MAP3K1rs=0
ThorssonTab$MAP3K1rs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"MAP3K1"]))
ThorssonTab$KMT2Crs=0
ThorssonTab$KMT2Crs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"KMT2C"]))
ThorssonTab$PALB2rs=0
ThorssonTab$PALB2rs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"PALB2"]))
ThorssonTab$PIK3CArs=0
ThorssonTab$PIK3CArs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"PIK3CA"]))
ThorssonTab$PTENrs=0
ThorssonTab$PTENrs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"PTEN"]))
ThorssonTab$TP53rs=0
ThorssonTab$TP53rs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"TP53"]))
ThorssonTab$TTNrs=0
ThorssonTab$TTNrs[na.omit(lx2)]=(sign(lx1[-which(is.na(lx2)) ,"TTN"]))


TTab2=ThorssonTab[ ,c("Lymphocyte Infiltration Signature Score", mSearch)]
colnames(TTab2)[1]="Immune"

```



```{r}
## do a t.test or something for each case?

ThorssonTab2=ThorssonTab[which(ThorssonTab$`TCGA Subtype`%in%c("BRCA.LumA", "BRCA.LumB")), ]

DNames=c("Macrophage Regulation","Lymphocyte Infiltration Signature Score","IFN-gamma Response",
         "Wound Healing", "TGF-beta Response")
stIdx=mSearch
Map2=matrix(NA, ncol=length(stIdx), nrow=length(DNames))
Map1=Map2


for (i in 1:length(DNames)){
Map1[i, ]=sapply(1:length(stIdx), function(x) t.test(as.numeric(pull(ThorssonTab2, DNames[i]))~pull(ThorssonTab2, stIdx[x]))$p.value)
Map2[i, ]=sapply(1:length(stIdx), function(x) diff(t.test(as.numeric(pull(ThorssonTab2, DNames[i]))~pull(ThorssonTab2, stIdx[x]))$estimate))
}

colnames(Map1)=stIdx
rownames(Map1)=DNames
Map2[which(Map1>0.05, arr.ind=T)]=NA
Map1[which(Map1>0.05, arr.ind=T)]=NA

colnames(Map2)=stIdx
rownames(Map2)=DNames

#heatmap.2(Map1, Rowv = NA, Colv = NA, col=blues9[9:1], trace="none")
heatmap.2(Map2, Rowv = NA, Colv = NA, col=RdBu[11:1], trace="none", symbreaks = T, scale="none")
```

A similar analysis can be performed on "predicted neoantigen" vs not. Again, we focus on ER+ cases as these mutations are most common in this subtype and there is an increased immune signature in ER- cases.

```{r}
ThorssonTab2=ThorssonTab[which(ThorssonTab$`TCGA Subtype`%in%c("BRCA.LumA", "BRCA.LumB")), ]

DNames=c("Macrophage Regulation","Lymphocyte Infiltration Signature Score","IFN-gamma Response",
         "Wound Healing", "TGF-beta Response")

stIdx=colnames(ThorssonTab2)[grep("[A-Z0-9]n$", colnames(ThorssonTab2))]
Map2=matrix(NA, ncol=length(stIdx), nrow=length(DNames))
Map1=Map2

for (i in 1:length(DNames)){
Map1[i,]=sapply(1:length(stIdx), function(x) t.test(as.numeric(pull(ThorssonTab2, DNames[i]))~pull(ThorssonTab2, stIdx[x]))$p.value)
Map2[i, ]=sapply(1:length(stIdx), function(x) diff(t.test(as.numeric(pull(ThorssonTab2, DNames[i]))~pull(ThorssonTab2, stIdx[x]))$estimate))

}

Map1b=matrix(p.adjust(Map1, "BH"), nrow=5)

colnames(Map1)=stIdx
rownames(Map1)=DNames
Map2[which(Map1>0.05, arr.ind=T)]=NA
Map1[which(Map1>0.05, arr.ind=T)]=NA

colnames(Map2)=stIdx
rownames(Map2)=DNames

#heatmap.2(Map1, Rowv = NA, Colv = NA, col=blues9[9:1], trace="none")
heatmap.2(Map2, Rowv = NA, Colv = NA, col=RdBu[11:1], trace="none", symbreaks = T, scale="none", sepwidth=c(0.005, 0.005),  # width of the borders
         colsep = c(0:10), rowsep = c(0: 6), sepcolor='black')


OutputMap=cbind(AllGMat, t(Map2))
```

For each case create a linear model that takes into account subtype, stage, and whether the mutation contributes to each signature

```{r}

DNames=c("Macrophage Regulation","Lymphocyte Infiltration Signature Score","IFN-gamma Response",
         "Wound Healing", "TGF-beta Response", "CTL", "Activated", "Inhibitory", "Exhausted", "Naive")

RNameCoef=c("PIK3CAn", "TP53n", "CDH1n","GATA3n","MUC16n","MAP3K1n","KMT2Cn","PALB2n","PTENn","TTNn")

GLMouput=lapply(DNames, function(x) summary(glm(as.numeric(pull(ThorssonTab2, x))~Stage+AKT1n+PIK3CAn+TP53n+CDH1n+GATA3n+MUC16n+
                  MAP3K1n+KMT2Cn+PALB2n+PTENn+TTNn,data=ThorssonTab2)))

GLMmat=sapply(GLMouput, function(x) ifelse(coefficients(x)[ ,4]<0.1, coefficients(x)[ ,1], 0))

T2=GLMmat[-c(1:6),  ]
colnames(T2)=DNames

T2[which(T2>0.6, arr.ind = T)]=0.6

heatmap.2(T2[sort(rownames(T2)), ], col=RdBu[11:1], trace="none", scale="none", Rowv = NA)


# ))
# 
# LIsignature=glm(as.numeric(pull(ThorssonTab2, DNames[1]))~Stage+PIK3CAn+TP53n+CDH1n+GATA3n+MUC16n+
#                   MAP3K1n+KMT2Cn+PALB2n+PTENn+TTNn,data=ThorssonTab2)
# 
# mac_reg=glm(as.numeric(`Macrophage Regulation`)~Stage+PIK3CAn+TP53n+CDH1n+GATA3n+MUC16n+
#                   MAP3K1n+KMT2Cn+PALB2n+PTENn+TTNn,data=ThorssonTab2)
# 
# ifnRes=glm(as.numeric(`IFN-gamma Response`)~Stage+PIK3CAn+TP53n+CDH1n+GATA3n+MUC16n+
#                   MAP3K1n+KMT2Cn+PALB2n+PTENn+TTNn,data=ThorssonTab2)
# 
# whresponse=glm(as.numeric(`Wound Healing`)~Stage+PIK3CAn+TP53n+CDH1n+GATA3n+MUC16n+
#                   MAP3K1n+KMT2Cn+PALB2n+PTENn+TTNn,data=ThorssonTab2)
# 
# tgfresponse=glm(as.numeric(`TGF-beta Response`)~Stage+PIK3CAn+TP53n+CDH1n+GATA3n+MUC16n+
#                   MAP3K1n+KMT2Cn+PALB2n+PTENn+TTNn,data=ThorssonTab2)
# 
# CD8response=glm(as.numeric(`TIL Regional Fraction`)~Stage+PIK3CAn+TP53n+CDH1n+GATA3n+MUC16n+
#                 MAP3K1n+KMT2Cn+PALB2n+PTENn+TTNn,data=ThorssonTab2)
```

## DCIS set

Perform a similar analysis to assess the frequency in DCIS. In this section, use the Abba cohort as well as the DCIS samples in the recurrence cohort. Overall, a similar pattern is observed where the most common mutations/neoantigens are in PIk3CA and in TP53.

```{r}

## Abba Set
mmAbba=data.matrix(N2[match(mSearch, rownames(N2)), -1]) # this is ok
rsSNPAbba=mmAbba
rsSNPAbba[which(rsSNPAbba!=6, arr.ind=T)]=0

AbbaSumm=AbbaNeoMutSum[match(mSearch, rownames(AbbaNeoMutSum)), ]

AbbaSumMat=cbind(rowSums(sign(mmAbba)),
                 sapply(1:nrow(AbbaSumm), function(x) length(which(AbbaSumm[x, ]>2))),  rowSums(sign(rsSNPAbba)))
AbbaSumMat[which(is.na(AbbaSumMat), arr.ind=T)]=0

colnames(AbbaSumMat)=c("N.DCIS.Mut", "N.DCIS.Neo", "N.DCIS.rsSNP" )

##Recurrence set
xrowN=sapply( strsplit(rownames(MutCNVTabrecur), " ) "), function(x) x[length(x)])
ReMut=MutCNVTabrecur[ match(mSearch, xrowN),seq(1, 12, by=2)]
ReMut[which(ReMut==2, arr.ind=T)]=0
ReMut[which(ReMut==(-2), arr.ind=T)]=0

NeoRec2=recNeo[match(mSearch, rownames(recNeo)), grep("DCIS", colnames(recNeo))]

RSumMat=cbind(rowSums(abs(ReMut), na.rm=T),rowSums(sign(abs(NeoRec2)), na.rm=T),
              sapply(1:nrow(ReMut), function(x) length(which(ReMut[x, ]==1))))


NMat=AbbaSumMat+RSumMat

rownames(NMat)=mSearch

## write this table to file, and write down the key mutation sites?
#write.csv(AbbaSumMat, file="rslt/Routput/DCIS_mutation_frequency.csv")


colnames(NMat)=c("nPat", "nRSNP", "nNeo")
NMat=t(t(NMat)/c(29+6, 29+6, 29+6))

amelt=melt(NMat)
ggplot(amelt, aes(x=Var1, y=value, fill=Var2))+geom_bar(stat="identity", position='dodge')+theme_bw()

```

Perform Proportion test to see if the frequency is different in DCIS and IDC

```{r}
OutputMap=cbind(OutputMap, AbbaSumMat)
OutputMap
AbbaSumMat

Ptest=sapply(1:nrow(OutputMap), function(x) prop.test( c(AbbaSumMat[x, 2], OutputMap[x, 11]), c(35, 977))$p.value)
names(Ptest)=rownames(OutputMap)
Ptest
```

write to file as Table S5

```{r}

write.csv(OutputMap, file="rslt/TableS5_DCIS_mutation_frequency.csv")
```

Test differences in gene expression or signatures between mutated vs not:

```{r}
FinMat=AbbaNeoMutSum[na.omit(match(mSearch, rownames(AbbaNeoMutSum))), ]

idxV2=match(colnames(FinMat), colnames(Imm2))

DCIStestMat=matrix(NA, ncol=nrow(FinMat), nrow=nrow(Imm2))

for (i in 1:nrow(FinMat)){
  for (j in 1:nrow(Imm2)){
    DCIStestMat[j, i]=tryCatch({t.test(Imm2[j , na.omit(idxV2)]~ifelse(FinMat[i, -which(is.na(idxV2))]>0, 1, 0))$p.value}, error=function(e) {NA})
  }
}

rownames(DCIStestMat)=rownames(Imm2)
colnames(DCIStestMat)=rownames(FinMat)
```

Also try a regression model

```{r}
FinMat2=FinMat
FinMat2[which(FinMat2<3, arr.ind = T)]=0
FinMat2=sign(FinMat2[ ,-which(is.na(idxV2))])

MatB=Imm2[ , na.omit(idxV2)]

df.B=cbind(t(FinMat2), t(MatB))
df.B=data.frame(df.B)

DNames=colnames(df.B)[7:19]

GLMouput=lapply(DNames, function(x) summary(glm(as.numeric(pull(df.B, x))~factor(GATA3)+factor(PIK3CA)+factor(TP53)+factor(KMT2C)+factor(MUC16)+factor(TTN),data=df.B)))

```